src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ValuesSources/WeaklyCachedRecoverableValueSource.cs

```diff
@@ -45,118 +66,118 @@ public WeaklyCachedRecoverableValueSource(ValueSource<T> initialValue)
+        /// <summary>
+        /// Kicks off the work to save this instance to secondary storage at some point in the future.  Once that save
+        /// occurs successfully, we will drop our cached data and return values from that storage instead.
+        /// </summary>
+        private void UpdateWeakReferenceAndEnqueueSaveTask_NoLock(T instance)
         {
-            saveTask.SafeContinueWith(t =>
-            {
-                using (Gate.DisposableWait(CancellationToken.None))
-                {
-                    // Only assume the instance is saved if the saveTask completed successfully. If the save did not
-                    // complete, we can still rely on a constant value source to provide the instance.
-                    _recoverySource = saveTask.Status == TaskStatus.RanToCompletion
-                        ? new AsyncLazy<T>(RecoverAsync, Recover, cacheResult: false)
-                        : new AsyncLazy<T>(instance);
-
-                    // Need to keep instance alive until recovery source is updated.
-                    GC.KeepAlive(instance);
-                }
-            }, TaskScheduler.Default);
-        }
+            Contract.ThrowIfTrue(Gate.CurrentCount != 0);
 
-        private Task? EnsureInstanceIsSavedAsync(T instance)
-        {
-            if (_weakReference == null)
-            {
-                _weakReference = new WeakReference<T>(instance);
-            }
-            else
-            {
-                _weakReference.SetTarget(instance);
-            }
+            _weakReference ??= new WeakReference<T>(instance);
+            _weakReference.SetTarget(instance);
 
+            // Ensure we only save once.
             if (!_saved)
             {
                 _saved = true;
                 using (s_taskGuard.DisposableWait())
                 {
-                    // force all save tasks to be in sequence so we don't hog all the threads
-                    s_latestTask = s_latestTask.SafeContinueWithFromAsync(t =>
-                         SaveAsync(instance, CancellationToken.None), CancellationToken.None, TaskScheduler.Default);
-                    return s_latestTask;
+                    // force all save tasks to be in sequence so we don't hog all the threads.
+                    s_latestTask = SaveAndResetInitialValue(s_latestTask);
                 }
             }
 
-#pragma warning disable VSTHRD114 // Avoid returning a null Task (False positive: https://github.com/microsoft/vs-threading/issues/637)
-            return null;
-#pragma warning restore VSTHRD114 // Avoid returning a null Task
+            return;
+
+            async Task SaveAndResetInitialValue(Task previousTask)
+            {
+                // First wait for the prior task in the chain to be done.  Ignore all errors from prior tasks.  They
+                // do not affect if we run or not.
+                await previousTask.NoThrowAwaitableInternal(captureContext: false);
```

Comment:
I was about to ask why we aren't just using the ContinueWith here rather than doing this manually and then noticed that's the old code. What was the specific concern here?

Reply:
Personally, i just find thsi code cleaner and easier to reason about.  I can switch to ContinueWith if you like, i just find it a bit more confusing as to what's going on.